src/beast/beast/container/vectum.h

```diff
@@ -0,0 +1,642 @@
+    {
+        return ptr_ != rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator< (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ < rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator> (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ > rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator<= (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ <= rhs.ptr_;
+    }
+
+    template <bool maybe_const>
+    bool
+    operator>= (basic_iterator<
+        maybe_const> const& rhs) const
+    {
+        return ptr_ >= rhs.ptr_;
+    }
+
+    basic_iterator&
+    operator++()
+    {
+        ptr_ = increment(1);
+        return *this;
+    }
+
+    basic_iterator
+    operator++ (int) const
+    {
+        basic_iterator _(*this);
+        ++(*this);
+        return _;
+    }
+
+    basic_iterator&
+    operator--()
+    {
+        ptr_ = increment(-1);
+        return *this;
+    }
+
+    basic_iterator
+    operator-- (int)
+    {
+        basic_iterator _(*this);
+        --(*this);
+        return _;
+    }
+
+    basic_iterator&
+    operator+= (std::ptrdiff_t n)
+    {
+        ptr_ = increment(n);
+        return *this;
+    }
+
+    basic_iterator&
+    operator-= (std::ptrdiff_t n)
+    {
+        ptr_ = increment(-n);
+        return *this;
+    }
+
+    friend
+    basic_iterator
+    operator+ (std::ptrdiff_t n,
+        basic_iterator const& rhs)
+    {
+        basic_iterator _(rhs);
+        _ += n;
+        return _;
+    }
+
+    friend
+    basic_iterator
+    operator+ (basic_iterator const& lhs,
+        std::ptrdiff_t n)
+    {
+        return basic_iterator(
+            lhs.increment(n), lhs.aligned_size_);
+    }
+
+    friend
+    basic_iterator
+    operator- (basic_iterator const& lhs,
+        std::ptrdiff_t n)
+    {
+        return basic_iterator(
+            lhs.increment(-n), lhs.aligned_size_);
+    }
+
+    friend
+    std::ptrdiff_t
+    operator- (basic_iterator const& lhs,
+                basic_iterator const& rhs)
+    {
+        assert(lhs.aligned_size_ == rhs.aligned_size_);
+        return
+            (reinterpret_cast<byte_t>(lhs.ptr_) -
+                reinterpret_cast<byte_t>(rhs.ptr_)) /
+                    lhs.aligned_size_;
+    }
+
+private:
+    friend class vectum;
+
+    basic_iterator (ptr_t ptr,
+            std::size_t aligned_size)
+        : ptr_ (ptr)
+        , aligned_size_ (aligned_size)
+    {
+    }
+
+    ptr_t
+    increment (std::ptrdiff_t n) const
+    {
+        return reinterpret_cast<ptr_t> (
+            (reinterpret_cast<byte_t>(ptr_) +
+                n * aligned_size_));
+    }
+};
+
+//------------------------------------------------------------------------------
+
+template <class T>
+vectum<T>::vectum (size_type value_size)
+    : value_size_ (value_size)
+    , aligned_size_ (((value_size +
```

Comment:
Have you tried this for many types `T`?

I'm unable to convince myself that these next four lines are accurate without actually using a pencil and paper...  Consider extracting out the logic as a little template function.

I have two candidates here - note that the second one save one multiplication by using `%` as well as being clearer IMHO:

    template <class T>
    std::size_t vecSize(size_t size)
    {
        auto align = std::alignment_of<T>::value;
        return ((size + align - 1) / align) * align;
    };

    template <class T>
    std::size_t vecSize(size_t size)
    {
        auto align = std::alignment_of<T>::value;
        auto full = size + align - 1;
        return full - (full % align);
    };